### PR TITLE
Throw if the genesis block in the storage doesn't match the configured genesis block

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
@Geod24 in general I don't like the way the code works. The `setGenesis(new) scope (exit) setGenesis(old)` reminds me a lot about `enableGagging` in DMD. Just the use of globals seems bad..

And in retrospect having the genesis block be defined in a config file is error-prone. If the file is corrupt, or if a user copy-pastes a configuration from some malicious website, they may end up on the wrong chain.

I'll have to think about a better alternative to `setGenesisBlock`..

### CODE REVIEW

File: source/agora/node/Ledger.d [hunk 1]
```diff
@@ -1346,3 +1352,50 @@ unittest
...
+
+    // hex => bin
+    auto block_bytes = gen_block_hex.chunks(2).map!(
+        twoDigits => twoDigits.parse!ubyte(16)).array();
+    immutable new_gen_block = block_bytes.deserializeFull!(immutable(Block));
```
Comment (MichaelKim20) [inline source/agora/node/Ledger.d:1359]: It's okay as it is, but if you add it, others will understand it better.

```suggestion
    immutable new_gen_block = block_bytes.deserializeFull!(immutable(Block));
    assert(new_gen_block != GenesisBlock);
```

**Review by MichaelKim20 — APPROVED** [review]
It's all good. I made a suggestion that is not serious.

**Comment by Geod24** [comment]
So we discussed this IRL with @AndrejMitrovic and talked about providing CLI switches instead. Where are we on this ?

**Comment by AndrejMitrovic** [comment]
See the comments in #821

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -1346,3 +1352,50 @@ unittest
...
+    auto utxo_set = new UTXOSet(":memory:");
+    scope (exit) utxo_set.shutdown();
+    auto config = new Config();
+    auto enroll_man = new EnrollmentManager(":memory:", gen_key_pair);
+    scope (exit) enroll_man.shutdown();
+    config.node.is_validator = true;
+
+    try
+    {
+        scope ledger = new Ledger(pool, utxo_set, storage, enroll_man, config.node);
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: Can you replace this with a `TestLedger` ?

File: source/agora/node/Ledger.d [hunk 3]
```diff
@@ -1346,3 +1352,50 @@ unittest
...
+    {
+        assert(ex.msg == "Genesis block loaded from disk is different to the one in the config file");
+    }
+
+    auto old_gen = &GenesisBlock();
+    setGenesisBlock(&new_gen_block);
+    scope (exit) setGenesisBlock(old_gen);  // must reset it for other tests
+
+    // will no longer fail
+    scope ledger = new Ledger(pool, utxo_set, storage, enroll_man, config.node);
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: Ditto `TestLedger`

File: source/agora/node/Ledger.d [hunk 4]
```diff
@@ -1346,3 +1352,50 @@ unittest
     genBlockTransactionsFreeze(1, TxType.Payment);
     assert(ledger.getBlockHeight() == 11 + 2016);
 }
+
+/// test that loading wrong genesis block throws
+unittest
+{
+    import agora.common.crypto.Key;
+    import agora.common.Types;
+    import agora.common.Hash;
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: Those three imports are already present at module level

File: source/agora/node/Ledger.d [hunk 5]
```diff
@@ -1346,3 +1352,50 @@ unittest
...
+    import agora.common.Serializer;
+    import std.conv;
+
+    auto gen_key_pair = getGenesisKeyPair();
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: This is gone

File: source/agora/node/Ledger.d [hunk 6]
```diff
@@ -1346,3 +1352,50 @@ unittest
...
+
+    // hex => bin
+    auto block_bytes = gen_block_hex.chunks(2).map!(
+        twoDigits => twoDigits.parse!ubyte(16)).array();
+    immutable new_gen_block = block_bytes.deserializeFull!(immutable(Block));
+    auto storage = new MemBlockStorage();
+    storage.saveBlock(new_gen_block);  // stores it as the genesis block
+
+    auto pool = new TransactionPool(":memory:");
+    scope(exit) pool.shutdown();
```
Comment (Geod24) [inline source/agora/node/Ledger.d]:
```suggestion
    scope (exit) pool.shutdown();
```

For consistency

**Review by Geod24 — COMMENTED** [review]
So, even with the approach of splitting the Genesis block, we still need this. Sorry for not seeing this earlier. Can you rebase / address review ?

File: source/agora/node/Ledger.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: I don't think so, I need to pass the MemBlockStorage loaded with the wrong block to the constructor of Ledger.

Comment (Geod24) [reply source/agora/node/Ledger.d]: `TestLedger` accepts an array of `Block` as second argument which will then be loaded by the `MemBlockStorage`.

**Comment by AndrejMitrovic** [comment]
Updated, using TestLedger now.

**Review by Geod24 — APPROVED** [review]
Noice

**Pull request merged**